### ISSUE TITLE
[haxe] Empty references when initially loading a .hxproj

### ISSUE THREAD

**Issue by Gama11** [fdorg/flashdevelop#1047]
This one is quite annoying. When loading a `.hxproj` file for the first time (cloned from a git repo like flixel-demos), the references node of the project manager is usually empty. Loading the project a second time fixes the problem, and it can't be reproduced anymore after that even when restarting FlashDevelop.

This isn't just a visual issue in the project manager, it also means that auto completion, goto declaration, QuickNavigate's type explorer etc aren't working properly.

![](http://i.imgur.com/fBjW97C.gif)

[(`.hxproj` file from the example)](https://github.com/HaxeFlixel/flixel-demos/blob/dev/Flixel%20Features/Colors/Colors.hxproj)

My guess is that this is caused by invalid / outdated `<classpaths>` entries - even if paths like `C:\HaxeToolkit\haxe\lib\lime\2,9,0` are often the same across computers, the version will change frequently and quickly become outdated. According to a diff, that and `<option directives>` / `<option additional>` are the only things that have changed.

**Comment by elsassph** [comment]
Correct, I noticed that too. Pressing the 'Refresh' icon fixes it also but it should be fixed.

**Comment by Gama11** [comment]
Seems like a threading / timing issue. `BroadcastProjectInfo()` in `SetActiveProject()` leads to the update of `HaxeProject#Classpaths` via `ExternalToolchain.UpdateProject()`. Since this happens async, the calls immediately after (responsible for telling ASCompletion what classpaths / refreshing the tree nodes) will still be using the old (in this case invalid) classpaths. 

```haxe
BroadcastProjectInfo(project);
projectActions.UpdateASCompletion(MainForm, project);

// ui
pluginUI.SetProject(project);
```

**Comment by elsassph** [comment]
Yes there's a kind of chicken-and-egg problem when opening a project. Openfl requires to run an operation depending on the project being loaded, which then in turn can dirty the project information which would have to be re-dispatched. This leads to awkward code to avoid infinite update loops.

Maybe the solution is to add a new clear steps like "preprocess" (when opening) and "validate" (when modifying project properties).
